[RFR]: Async i18n messages proposal

tkvw opened the pull request:
This PR allows i18n messages to be loaded async and allows to change the locale through a query param. 
Usage: 
```
<Admin
messages={{
fr: import('./i18n/fr.js').then(m=>m.default)
}}
```
![2017-12-11_13-27-05](https://user-images.githubusercontent.com/621098/33831246-90e0dec0-de77-11e7-8101-5387f0ac0cf4.gif)

djhi commented:
IMO, it would be better to add/change the current locale actions so that they may include messages. This way, it would be possible to load them on demand and from any source (file, db, API, whatever).

tkvw commented:
@djhi : I changed the implementation to what you've suggested. I had some issues with my initial solution and this implementation is cleaner.

tkvw commented:
Moved list query issue to separate PR https://github.com/marmelab/admin-on-rest/pull/1326 and rebased.

Phocea commented:
A way to force reload of the locale without reloading the app would also be great. Then we could administer and reload locale via react-admin :)

fzaninotto commented on packages/ra-example/app.js:
```diff
@@ -21,13 +19,9 @@ import {
-    fr: { ...frenchMessages, ...customMessages.fr },
-    en: { ...englishMessages, ...customMessages.en },
-};
+import englishMessages from './i18n/en';
```
you probably forgot to update the English messages to use the custom messages, like in French.

fzaninotto commented on packages/react-admin/src/Admin.js:
```diff
@@ -44,7 +45,11 @@ const Admin = ({
-            [crudSaga(dataProvider, authClient), ...customSagas].map(fork)
+            [
+                crudSaga(dataProvider, authClient),
+                i18nSaga({ locale, messages, enableLocaleQuerySwitcher }),
```
I'd prefer that you put the `i18nSaga` in the `crudSaga`, to avoid composing sagas here. Remember, the code of the `Admin` component becomes the code to duplicate for people who want to build a custom app. The less work they have, the better.

fzaninotto commented on packages/react-admin/src/sideEffect/saga/i18nSaga.js:
```diff
@@ -0,0 +1,68 @@
+    return function* watchLocaleChanges() {
+        yield all([
+            takeLatest(CHANGE_LOCALE, handleChangeLocaleAction),
+            put(changeLocale(initialLocale)),
```
won't this create a rerender each time the app renders for the first time?

fzaninotto commented on packages/react-admin/src/sideEffect/saga/i18nSaga.js:
```diff
@@ -0,0 +1,68 @@
+
+export default ({
+    locale: initialLocale,
+    messages,
```
Storing the messages in the saga feels weird to me. I can't put the finger on why - probably because usually, sagas don't persist anything. storing stuff is the store's job.

fzaninotto commented on packages/react-admin/src/actions/localeActions.js:
```diff
@@ -4,3 +4,12 @@ export const changeLocale = locale => ({
+export const changeMessages = (locale, messages) => ({
+    type: CHANGE_MESSAGES,
+    payload: {
+        locale,
```
why does this action take the `locale` as parameter?

tkvw commented:
Changed as requested, introduced an `i18nDataprovider` and moved locale change using query param to user domain. I've included it in the example-app as a custom saga.

djhi commented on packages/react-admin/src/Admin.js:
```diff
@@ -29,10 +29,10 @@ const Admin = ({
     menu = Menu,
     catchAll,
     dataProvider,
+    i18nProvider,
```
Can we set a default `i18nProvider` which would load messages as before ?

tkvw replied:
```locale => this.props.messages[locale]``` ?

djhi commented on docs/Translation.md:
```diff
@@ -343,3 +379,31 @@ translate('not_yet_translated', { _: 'Default translation' })
+
+## Using asynchronous messages loading 
+
+By default React Admin loads all configured lanuages and languagebundles on startup. If you have very large language files, you can load messagebundles asynchronous. 
```
s/languagebundles/their messages/

djhi replied:
s/messagebundles/messages/

djhi replied:
s/asynchronous/asynchronously/

djhi commented on docs/Translation.md:
```diff
@@ -343,3 +379,31 @@ translate('not_yet_translated', { _: 'Default translation' })
+## Using asynchronous messages loading 
+
+By default React Admin loads all configured lanuages and languagebundles on startup. If you have very large language files, you can load messagebundles asynchronous. 
+React-admin uses an `i18nProvider` to load messages. A custom dataprovider can be supplied to the `Admin` component by using the `i18nProvider` prop. 
```
s/dataprovider/i18nProvider/

djhi commented on docs/Translation.md:
````diff
@@ -343,3 +379,31 @@ translate('not_yet_translated', { _: 'Default translation' })
+By default React Admin loads all configured lanuages and languagebundles on startup. If you have very large language files, you can load messagebundles asynchronous. 
+React-admin uses an `i18nProvider` to load messages. A custom dataprovider can be supplied to the `Admin` component by using the `i18nProvider` prop. 
+The signature of the `i18nProvider` is: 
+```jsx
````
Please add a new empty line above. Otherwise, it will not render correctly.

Screenshot:

![screen](https://i.imgur.com/hjMaQJL.png)

djhi commented on docs/Translation.md:
````diff
@@ -343,3 +379,31 @@ translate('not_yet_translated', { _: 'Default translation' })
+The promise should resolve to a messages object. 
+
+An example asynchronous `i18nProvider` is: 
+```jsx
````
Same here

djhi commented on docs/Translation.md:
````diff
@@ -343,3 +379,31 @@ translate('not_yet_translated', { _: 'Default translation' })
+```
+The promise should resolve to a messages object. 
+
+An example asynchronous `i18nProvider` is: 
````
s/is/would be/

djhi commented on docs/Translation.md:
````diff
@@ -343,3 +379,31 @@ translate('not_yet_translated', { _: 'Default translation' })
+export default asyncI18nProvider;
+```
+
+Like in this provider, it makes sense to include the primary language of your app synchronous.
````
s/synchronous/synchronously/

djhi commented on packages/react-admin/src/reducer/i18n/index.js:
```diff
@@ -0,0 +1,16 @@
+        locale: locale(initialLocale),
+        messages,
+        loading,
+        requested,
```
I don't understand this `requested` reducer. Where is it used ?

tkvw replied:
It's not used in `react-admin`, so maybe it shouldn't be there. It's there so an user can query the locale being loaded. `{locale:'en',requested:null}` -> `{type:CHANGE_LOCALE,payload:'nl'}` -> `{locale:'en',requested:'nl'}` ->possibility to show ui loading `nl` -> `{type:CHANGE_LOCALE_SUCCESS}` -> `{locale:'nl',requested:null}`

djhi replied:
I think this should stay in userland then

djhi commented on docs/Translation.md:
```diff
@@ -343,3 +379,34 @@ translate('not_yet_translated', { _: 'Default translation' })
+
+## Using asynchronous messages loading 
+
+By default React Admin loads all configured lanuages and their messages on startup. If you have very large language files, you can load messages asynchronously. 
```
s/lanuages/languages/

djhi commented:
Well, it seems we have a problem with our e2e tests on travis only. I can confirm the tests pass locally for your PR. I'm investigating

djhi commented:
We finally fixed the e2e tests issue. Can you please rebase your PR ?

fzaninotto commented:
I don't understand why the `yarn.lock` was modified, since there is no modification of the `package.json`.... Can you remove the change from that file (that would remove the need for a rebase)?

tkvw commented:
My bad..

fzaninotto commented on packages/react-admin/src/Admin.js:
```diff
@@ -29,10 +29,10 @@ const Admin = ({
     menu = Menu,
     catchAll,
     dataProvider,
+    i18nProvider = locale => this.props.messages[locale],
```
please extract this function to a `defaultI18nProvider` to avoid redefining a new function (and therefore failing the prop identity check) whenever the `<Admin>` component is rerendered.

fzaninotto commented on packages/react-admin/src/Admin.js:
```diff
@@ -63,7 +66,7 @@ const Admin = ({
     return (
         <Provider store={store}>
-            <TranslationProvider messages={messages}>
+            <TranslationProvider>
```
you need to update the `CustomApp` documentation, too.

fzaninotto replied:
Still required

fzaninotto commented on packages/react-admin/src/i18n/TranslationProvider.js:
```diff
@@ -3,8 +3,8 @@ import PropTypes from 'prop-types';
 import { compose, withContext } from 'recompose';
-
 import defaultMessages from 'ra-language-english';
+import defaultsDeep from 'lodash/defaultsDeep';
```
Use `lodash.defaultsdeep` instead (already a dependency)

fzaninotto commented on packages/react-admin/src/sideEffect/saga/i18n.js:
```diff
@@ -0,0 +1,33 @@
+        yield all([
+            takeLatest(CHANGE_LOCALE, loadMessages),
+            takeLatest(USER_LOGOUT, loadInitialMessages),
+            call(loadInitialMessages),
```
I'm afraid that this one will cause an initial double render of the `<Admin>` children. Can you check that it doesn't?

tkvw replied:
Yes I had the same concern, my observations are that it does when the messages are loaded async, but synchronous I don't see this effect. In my custom app I have a loader which `races` the `CHANGE_LOCALE_SUCCESS` and `CHANGE_LOCALE_FAILURE` before displaying `react-admin`, maybe this should be added?

fzaninotto replied:
Let's summarize the problems: 

1. The loading of the messages can be async to avoid loading all messages at startup
2. We don't want to block the rendering of the app while it waits for an async message fetch
3. But the app requires a default locale, loaded synchronously, for the first render
4. The locale can depend on the user (a custom `authSaga` can dispatch a `CHANGE_LOCALE` upon login).

The solution, IMO, is to implement two actions in the `i18nProvider`:

- `GET_DEFAULT_MESSAGES`: must be synchronous, is called by the `<Admin>` component, then passed to the i18n *reducer* to initialize the default i18n state
- `CHANGE_LOCALE`: can be asynchronous, and is called by the i18n *sagas*

That way, no need to reset messages upon logout, because the state has a default set of i18n messages.

The way I see it, the i18nProvider could still be a simple function taking the locale as only argument, and returning a promise for messages (if the locale argument is defined), or directly messages (if no locale is passed). Or, it could be a more structured function, similar to the `dataProvider`, taking type and params, as follows:

```js
const i18nProvider = (messages, defaultMessages) => (type, params) => {
    if (type === 'GET_DEFAULT_MESSAGES') {
        return defaultMessages;
    }
    if (type === 'CHANGE_LOCALE') {
        return messages[params.locale];
    }
    throw new Error('Undefined action type', type); 
}
```

Having to pass both a `messages` and a `defaultMessages` props to the `<Admin>` component will be cumbersome, so I guess we can take advantage of the BC break in 2.0 and ask only for one parameter: the `i18nProvider`. The default `i18nProvider` would simply be:

```js
import englishMessages from 'ra-language-english'
const i18nProvider = (type, params) => englishMessages;
```

What do you think?

fzaninotto commented on packages/react-admin/src/sideEffect/saga/i18n.js:
```diff
@@ -0,0 +1,33 @@
+    return function*() {
+        yield all([
+            takeLatest(CHANGE_LOCALE, loadMessages),
+            takeLatest(USER_LOGOUT, loadInitialMessages),
```
What's the rationale of doing it on user logout?

tkvw replied:
the messages are removed from the store using the `resettableAppReducer`, so they should be reloaded on user logout. Also the user locale can be different from the site default locale.

fzaninotto replied:
That means, IMO, that the *initial state* of the i18n reducer is wrong.

tkvw replied:
I see what you're saying, but my problem is with the how. The messages aren't known on initial render of `Admin`, not until the `i18n` saga finishes loading them. I can add loading state to `Admin`, but that complicates `Admin` more, any suggestions?

tkvw commented:
Thanks for your feedback! Very nice way of solving this. I've refactored the code to reflect the changes, I've changed `i18nProvider` to `messagesProvider` which is a better name in my opinion. Let me know what you think.

fzaninotto commented on docs/Translation.md:
```diff
@@ -343,3 +389,52 @@ translate('not_yet_translated', { _: 'Default translation' })
+            throw new Error(`Undefined messagesProvider action type ${type}`);
+    }
+};
+export default createMessagesProvider(undefined, null);
```
bad export

fzaninotto commented on examples/simple/sagas.js:
```diff
@@ -0,0 +1,28 @@
+
+const localeParam = 'locale';
+
+function* handleChangeLocaleAction({ payload: { search } }) {
```
I'm not fan of that use case - we shouldn't encourage messing up with the url. Changing the locale should be a function exposed in the UI. Besides, to illustrate how a custom saga can change the messages, why don't you illustrate the use of cookies for the locale?

fzaninotto commented on docs/Translation.md:
```diff
@@ -12,27 +12,37 @@ The react-admin interface uses English as the default language. But it also supp
+Provides messages for a specific locale. React-admin ships with a `createMessagesProvider` factory to easily create your own `messagesProvider`. 
+`createMessagesProvider` requires a single argument `messageBundles` and allows an optional `defaultMessageBundle`. 
+* `messageBundles`: required, an object with a language pack per locale. These language packs can be resolved async by returning a Promise.
+* `defaultMessageBundle`: optional, a default language pack where the application starts with. The `defaultMessageBundle` must always be resolved synchronous and defaults to the messages from `ra-language-english`.
```
no, it's not optional, on the contrary: it's THE required parameter.

fzaninotto commented on docs/Translation.md:
```diff
@@ -12,27 +12,37 @@ The react-admin interface uses English as the default language. But it also supp
+## `messagesProvider`
+
+Provides messages for a specific locale. React-admin ships with a `createMessagesProvider` factory to easily create your own `messagesProvider`. 
+`createMessagesProvider` requires a single argument `messageBundles` and allows an optional `defaultMessageBundle`. 
```
I'm not fan of this `createMessagesProvider`. It obfuscates the logic behind the provider, and it introduces a second way of doing i18n - which is hard to explain. 

I think you can document how to pass custom messages simply using the following example:

```js
const messages = {
  en: englishMessages,
  fr: frenchMessages,
};
const i18nProvider = (type, params) => {
    if (type === 'GET_DEFAULT_MESSAGES') {
        return englishMessages;
    }
    if (type === 'CHANGE_LOCALE') {
        return messages[params.locale];
    }
    throw new Error('Undefined action type', type); 
}
```

fzaninotto commented on docs/Translation.md:
````diff
@@ -125,6 +135,42 @@ class LocaleSwitcher extends Component {
 export default connect(undefined, { changeLocale: changeLocaleAction })(LocaleSwitcher);
 ```
 
+## Changing the locale using the browser URL 
````
Not the best illustration for a custom saga IMO (see other comment on that code snippet)

fzaninotto commented on docs/Translation.md:
```diff
@@ -12,27 +12,37 @@ The react-admin interface uses English as the default language. But it also supp
 
 - a `locale` prop expecting a string ('en', 'fr', etc), and
-- a `messages` prop, expecting a dictionary object.
+- a `messagesProvider` prop, expecting an `messagesProvider` as explained below.
```
I liked the `i18nProvider` name. The term *messages* isn't specific enough IMO (it can relate to notifications in many people's mind).

fzaninotto commented on packages/react-admin/src/reducer/i18n/index.js:
```diff
@@ -0,0 +1,14 @@
+import loading from './loading';
+
+export default (initialLocale, defaultMessages) => {
+    return combineReducers({
```
no need to `return` here, use the braceless function form.

fzaninotto commented on packages/react-admin/src/reducer/index.js:
```diff
@@ -5,16 +5,17 @@ import admin, {
     combineReducers({
         admin,
-        locale: localeReducer(locale),
+        i18n: i18n(locale, messages),
```
the function should be named `i18nReducer` for consistency.

fzaninotto commented on packages/react-admin/src/sideEffect/saga/crudSaga.js:
```diff
@@ -3,13 +3,15 @@ import auth from './auth';
  * @param {Object} dataProvider A Data Provider function
  */
-export default (dataProvider, authClient) =>
+export default (dataProvider, authClient, messagesProvider) =>
```
i18nProvider

tkvw commented:
Reverted back to `i18nProvider`, updated the docs and updated requested changes.

fzaninotto commented on docs/Translation.md:
```diff
@@ -12,27 +12,60 @@ The react-admin interface uses English as the default language. But it also supp
-- a `messages` prop, expecting a dictionary object.
+- a `i18nProvider` prop, expecting an `i18nProvider` as explained below.
 
-React-admin ships with the English locale by default. If you want to use another locale, you'll have to install a third-party package. For instance, to change the interface to French, install the `ra-language-french` npm package, then configure the `<Admin>` component as follows:
```
React-admin can still work without provider and default to English, right?

tkvw replied:
Yes, the default provider loads the english messages. Not sure if this should be changed to an empty messages object, because the same bundle is loaded in the `TranslationProvider` as default messages.  Having no messages in the defaultProvider feels weird, let me know what you want.

tkvw replied:
Yeah, so I rephrased it a bit, probably should have left the `React-admin ships with the English locale by default.` sentence in there, will change.

fzaninotto commented on docs/Translation.md:
````diff
@@ -12,27 +12,60 @@ The react-admin interface uses English as the default language. But it also supp
 
 ```jsx
 import React from 'react';
 import { Admin, Resource, resolveBrowserLocale } from 'react-admin';
````
I'd import the `GET_DEFAULT_MESSAGES` and `GET_LOCALE_MESSAGES` constants, too.

djhi commented on UPGRADE.md:
```diff
@@ -63,6 +63,45 @@ import { jsonServerRestClient } from 'admin-on-rest';
+  en: () => englishMessages,
+  fr: () => import('ra-language-french').then( bundle => bundle.default ),
+};
+const i18nProvider = (type, params) => {
```
This is already an advanced usage scenario. I'd rather keep this example the most simple possible using the default i18nProvider to have the same behavior as before. Then I'll introduce what this new provider allows one to do in a separate paragraph (something like *Lazily loading a locale messages*)

Phocea commented:
I have a question. 
Since the proper loading of the application will depend on an async call to fetch the locale, should we use something like [PersistGate](https://github.com/rt2zz/redux-persist/blob/6547840c3c1971b6af602c5400fa58a7efa2540b/docs/PersistGate.md), to ensure the App is ready ?

More over, the whole logic of [redux persis](https://github.com/rt2zz/redux-persist) could be used and the locale added to the store ?

Locale aside, having the possibility to at least use some kind of redux persistence mechanism without moving to a custom app would be a great addition to react-admin.

tkvw commented:
Hello @Phocea : there is no async call to fetch the locale. This PR focuses at loading the messages of a locale async. The provider requires the initial application messages to be loaded synchronous, so the loading of the application can continue. You are allowed to trigger a `changeLocale` from a startup saga to setup the correct locale of the user and load the messages sync/async (you're free to implement whatever you want in the `i18nProvider`)

djhi commented on UPGRADE.md:
```diff
@@ -63,6 +63,45 @@ import { jsonServerRestClient } from 'admin-on-rest';
+};
+const i18nProvider = (type, params) => {
+    if (type === GET_DEFAULT_MESSAGES) {
+        return messages.en; // The default messages should always resolve synchronous!! 
```
s/synchronous/synchronously/

djhi commented on UPGRADE.md:
```diff
@@ -63,6 +63,45 @@ import { jsonServerRestClient } from 'admin-on-rest';
 
+## Introduced `i18nProvider` for loading of locale messages
+
+Previously you would supply `react-admin` with a `messages` prop object to allow alternative locales, this should be replaced with an `i18nProvider`. The new `i18nProvider` allows you to load the messages asynchronously. @See [`i18nProvider`](./Translation.md#i18nProvider) 
```
What's the `@` before `See` ?

tkvw replied:
jsdoc notation, shouldn't be there

fzaninotto commented on UPGRADE.md:
````diff
@@ -63,6 +63,45 @@ import { jsonServerRestClient } from 'admin-on-rest';
+```jsx
+// before
+import { enMessages } from 'admin-on-rest';
+const messages = { 'en': enMessages };
````
the two examples are not the same, since you're passing one language here and two in the second snippet. An upgrade guide should show how to achieve *the same thing* on the two versions.

fzaninotto commented on docs/CustomApp.md:
```diff
@@ -52,9 +52,17 @@ import { UserList, UserEdit, UserCreate } from './User';
+    if (type === 'CHANGE_LOCALE') {
+        return messages[params.locale];
+    }
+    throw new Error('Undefined action type', type); 
```
this won't work, the initial message is `GET_DEFAULT_MESSAGES` and it will be rejected. Besides, isn't it `GET_LOCALE_MESSAGES`?

fzaninotto commented:
@tkvw Thanks a lot for your contribution, it is much appreciated!

As you may have seen on this PR, managing contributions requires a lot of time on our part. In the near future, we want to focus on releasing the 2.0 rather than adding new features. If you want to give us a hand, please triage new issues/PRs, fix bugs, and write documentation.

tkvw commented:
@fzaninotto : np! Thanks for your inputs. I'll try to be more secure on my side the next time. Can you add labels 2.0.0 milestone labels to the PR's you want to include, so I can focus on them?

fzaninotto commented:
If you can start by reading every open issue and seeing if you can help reproduce / localize the reported bugs, that would be a great help/
